internal/charmstore/debug.go

```diff
@@ -14,60 +14,304 @@ import (
+type debugParams struct {
+	ServerParams
+	pool            *Pool
+	loopbackHandler http.Handler
 }
 
-// GET /debug/check.
-func debugCheck(checks map[string]func() error) http.Handler {
-	return router.HandleJSON(func(http.Header, *http.Request) (interface{}, error) {
-		n := len(checks)
-		type result struct {
-			name string
-			err  error
+type debugHandler struct {
+	router    *httprouter.Router
+	params    debugParams
+	checks    map[string]func() error
+	openPaths map[string]bool
+}
+
+func newDebugHandler(p debugParams) *debugHandler {
+	hnd := debugHandler{
+		router: httprouter.New(),
+		params: p,
+		checks: map[string]func() error{
+			"mongodb":       checkDB(p.pool.db.Database),
+			"elasticsearch": checkES(p.pool.es),
+		},
+		openPaths: map[string]bool{
+			"/debug/check":  true,
+			"/debug/info":   true,
+			"/debug/status": true,
+		},
+	}
+	for _, h := range router.ErrorToResp.Handlers(hnd.handle) {
+		hnd.router.Handle(h.Method, h.Path, h.Handle)
+	}
+	return &hnd
+}
+
+func (h *debugHandler) handle(p httprequest.Params) (*debugReqHandler, error) {
+	if err := h.authorized(p.Request); err != nil {
+		return nil, err
+	}
+	rh := debugReqHandler{
+		Handler: debugstatus.Handler{
+			Version: debugstatus.Version(appver.VersionInfo),
+			// Authorization has alread been checked so there is no need to do it again.
+			CheckPprofAllowed: func(*http.Request) error { return nil },
+		},
+		h: h,
+	}
+	rh.Handler.Check = rh.check
+	return &rh, nil
+}
+
+func (h *debugHandler) authorized(r *http.Request) error {
+	if h.openPaths[r.URL.Path] {
+		return nil
+	}
+	// TODO support other login methods
+	u, p, err := utils.ParseBasicAuthHeader(r.Header)
+	if err != nil {
+		return errgo.WithCausef(err, params.ErrUnauthorized, "")
+	}
+	if u != h.params.AuthUsername || p != h.params.AuthPassword {
+		return errgo.WithCausef(nil, params.ErrUnauthorized, "username or password mismatch")
+	}
+	return nil
+}
+
+func (h *debugHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// The charmstore router will have stripped "/debug" off the
+	// front of the path, put it back.
+	r.URL.Path = "/debug" + r.URL.Path
+	h.router.ServeHTTP(w, r)
+}
+
+type debugReqHandler struct {
+	debugstatus.Handler
+	h *debugHandler
+
+	// store is initialized lazily.
+	store_     *Store
+	storeError error
+}
+
+func (h *debugReqHandler) Close() error {
+	if h.store_ != nil {
+		h.store_.Close()
+	}
+	return nil
+}
+
+func (h *debugReqHandler) store() (*Store, error) {
+	if h.store_ == nil && h.storeError == nil {
+		h.store_, h.storeError = h.h.params.pool.RequestStore()
+	}
+	return h.store_, h.storeError
+}
+
+func (h *debugReqHandler) check() map[string]debugstatus.CheckResult {
+	checkers := []debugstatus.CheckerFunc{
+		debugstatus.ServerStartTime,
+		h.checkElasticSearch,
+	}
+	store, err := h.store()
+	if err != nil {
+		checkers = append(checkers, func() (string, debugstatus.CheckResult) {
+			return "store available", debugstatus.CheckResult{
```

Comment:
Yeah, but it's kinda strange to have a test entry that only exists when there's an error.